src/Storage/TSStorageManager.m

```diff
@@ -92,24 +110,52 @@ + (instancetype)sharedManager {
 - (instancetype)initDefault
 {
     self = [super init];
+    
+    if (![self tryToLoadDatabase]) {
+        // Failing to load the database is a catastrophic.
+        //
+        // The best we can try to do is to discard the current database
+        // and behave like a clean install.
+        
+        OWSAnalyticsCritical(@"Could not load database");
+    
+        // Try to reset app by deleting database.
+        [self deletePasswordFromKeychain];
+        [self deleteDatabaseFile];
+
+        if (![self tryToLoadDatabase]) {
+            OWSAnalyticsCritical(@"Could not load database (second attempt)");
+
+            [NSException raise:TSStorageManagerExceptionNameNoDatabase format:@"Failed to initialize database."];
+        }
+    }
+    
+    return self;
+}
 
+- (BOOL)tryToLoadDatabase {
+    
+    // We determine the database password first, since a side effect of
+    // this can be deleting any existing database file (if we're recovering
+    // from a corrupt keychain).
+    NSData *databasePassword = [self databasePassword];
```

Comment:
You know. THIS change made me curious, and I found out we were calling the cipher block which queries the keychain **25 times** in quick succession on startup. =/